pkg/cmd/admin/image/verify-signature.go

```diff
@@ -0,0 +1,373 @@
+		return kcmdutil.UsageError(cmd, "exactly one image must be specified")
+	}
+	o.InputImage = args[0]
+	var err error
+
+	// If --public-key is provided only this key will be used for verification and the
+	// .gnupg/pubring.gpg will be ignored.
+	if len(o.PublicKeyFilename) > 0 {
+		if o.PublicKey, err = ioutil.ReadFile(o.PublicKeyFilename); err != nil {
+			return err
+		}
+	}
+	if o.Client, _, err = f.Clients(); err != nil {
+		return err
+	}
+	// Only make this API call when we are sure we will be writing validation.
+	if o.Confirm && !o.Remove {
+		if me, err := o.Client.Users().Get("~"); err != nil {
+			return err
+		} else {
+			o.CurrentUser = me.Name
+		}
+	}
+
+	return nil
+}
+
+func (o VerifyImageSignatureOptions) Run() error {
+	img, err := o.Client.Images().Get(o.InputImage)
+	if err != nil {
+		return err
+	}
+	if len(img.Signatures) == 0 {
+		return fmt.Errorf("%s does not have any signature", img.Name)
+	}
+
+	policy, err := signature.DefaultPolicy(nil)
+	if err != nil {
+		return fmt.Errorf("Error reading signature verification policy: %v", err)
+	}
+	pc, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("Error preparing for signature verification: %v", err)
+	}
+	defer pc.Destroy()
+
+	for i, s := range img.Signatures {
+		// If --remove is specified, just erase the existing signature verification for all
+		// signatures.
+		// TODO: This should probably need to handle removal of a single signature.
+		if o.Remove {
+			o.clearSignatureVerificationStatus(&img.Signatures[i])
+			continue
+		}
+
+		// Verify the signature against the policy
+		signedBy, err := o.verifySignature(pc, img, s.Content)
+		if err != nil {
+			fmt.Fprintf(o.ErrOut, "error verifying %s signature %d: %v\n", o.InputImage, i, err)
+			o.clearSignatureVerificationStatus(&img.Signatures[i])
+			continue
+		}
+		fmt.Fprintf(o.Out, "%s signature %d is verified (signed by key: %q)\n", o.InputImage, i, signedBy)
+
+		now := unversioned.Now()
+		newConditions := []imageapi.SignatureCondition{
+			{
+				Type:               imageapi.SignatureTrusted,
+				Status:             kapi.ConditionTrue,
+				LastProbeTime:      now,
+				LastTransitionTime: now,
+				Reason:             "verified manually",
+				Message:            fmt.Sprintf("verified by user %q", o.CurrentUser),
+			},
+			// FIXME: This condition is required to be set for validation.
+			{
+				Type:               imageapi.SignatureForImage,
+				Status:             kapi.ConditionTrue,
+				LastProbeTime:      now,
+				LastTransitionTime: now,
+			},
+		}
+		img.Signatures[i].Conditions = newConditions
+		img.Signatures[i].IssuedBy = &imageapi.SignatureIssuer{}
+		// TODO: This should not be just a key id but a human-readable identity.
+		img.Signatures[i].IssuedBy.CommonName = signedBy
+	}
+
+	if o.Confirm {
+		_, err := o.Client.Images().Update(img)
+		return err
+	}
+	fmt.Fprintf(o.Out, "(add --confirm to record signature verification status to server)\n")
```

Comment:
@smarterclayton any suggestion on how to make it better? the intent is to not surprise the user by the fact that the verification is not recorded ...